File: src/year2024/day3/day3.py
#!/usr/bin/env python3
# [Day 3: Mull It Over](https://adventofcode.com/2024/day/3)

import re
import sys
from pathlib import Path

if verbose := "-v" in sys.argv:
    sys.argv.remove("-v")
if self_tests := "-T" in sys.argv:
    sys.argv.remove("-T")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"


def part1(data: str) -> int:
    return sum(int(m[1]) * int(m[2]) for m in re.finditer(r"mul\((\d+),(\d+)\)", data))


def part2(data: str) -> int:
    enabled = True
    total_sum = 0
    for m in re.finditer(r"mul\((\d+),(\d+)\)|don't\(\)|do\(\)", data):
        if m[0] == "do()":
            enabled = True
        elif m[0] == "don't()":
            enabled = False
        elif enabled:
            total_sum += int(m[1]) * int(m[2])

    return total_sum


if self_tests:
    assert part1(Path("sample_1.txt").read_text()) == 161
    assert part2(Path("sample_2.txt").read_text()) == 48

else:
    data = Path(filename).read_text().strip()
    print(part1(data))
    print(part2(data))

File: src/year2024/day3/day3.rs
//! [Day 3: Mull It Over](https://adventofcode.com/2024/day/3)

use regex::Regex;

/// Compute valid `mul()` operations.
/// if part2 is true, take care of `do()`/`don't()` statements.
fn calc(data: &str, part: u8) -> i32 {
    let mut enabled = true;
    let mut total_sum = 0;

    let pattern = Regex::new(r"mul\((\d+),(\d+)\)|do\(\)|don't\(\)").unwrap();

    for m in pattern.captures_iter(data) {
        if m.get(0).unwrap().as_str() == "do()" {
            enabled = true;
        } else if m.get(0).unwrap().as_str() == "don't()" {
            enabled = false;
        } else if enabled || part == 1 {
            let x = m[1].parse::<i32>().unwrap();
            let y = m[2].parse::<i32>().unwrap();
            total_sum += x * y;
        }
    }

    total_sum
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (i32, i32) {
    (calc(data, 1), calc(data, 2))
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE_1: &str = include_str!("sample_1.txt");
    const SAMPLE_2: &str = include_str!("sample_2.txt");

    #[test]
    fn test01() {
        assert_eq!(calc(SAMPLE_1, 1), 161);
        assert_eq!(calc(SAMPLE_1, 2), 161);
    }

    #[test]
    fn test02() {
        assert_eq!(calc(SAMPLE_2, 1), 161);
        assert_eq!(calc(SAMPLE_2, 2), 48);
    }
}
